# Career Exploration Kaggle Project

## The Data Science Workflow

In general, there are a few key steps to begin working with a dataset.

_First_, we need to understand what the dataset actually is about, and what we are trying to 
do with it. The key to this stage is understanding what each row of the dataset represents, as 
well as what each column indicates. You can read more about this dataset by looking at the 
dataset itself (https://www.kaggle.com/c/house-prices-advanced-regression-techniques/), or 
reading the __data dictionary__ in BLANK.

_Second_, before we can even do exploratory data analysis, we need to __clean our dataset__. 
It is very helpful to identify the size of the dataset, so we know how many samples we 
have. We also need to determine a consistent method of dealing with missing values, such as 
setting them to a value, removing the feature entirely, interpolating values, etc. Other 
crucial steps are separating into training and validation, as well as creating elementary 
data plots.

_Third_, we have __exploratory data analysis (EDA)__. The point of this phase is to inspect & 
visualize key relationships, trends, outliers, and issues with our data. By conducting 
EDA, we get a better sense of the underlying structure of our data, and which features are 
most important. Especially since we have 81 features, we would like to __select__ the features 
that are most important to our analysis before we begin modeling SalePrice. Once we have 
an idea about which features to use and how they relate to one another, our modeling stage 
will be more informed and robust.

_Fourth_, we have the modeling phase, consisting of __model selection__ and __model training__. Here, we select a predictive model to train on our features, and then actually train the model! In this first week, we will be using linear regression as a first-pass. In later weeks of the SUSA CX Kaggle Capstone project, we will be using more advanced models like 
__random forest__ and __neural networks__. Depending on the model we are using, it is important to verify the model's assumptions before fully pledging to that model. Additionally, we may use validation in this stage to select certain hyperparameters for our model selection.

_Finally_, we have the __model evaluation__ phase. Here, we compute a metric for our model's 
performance, usually by summing the squared errors of the model's predictions on the test 
set. This stage allows us to effectively compare various data cleaning and modeling 
selection decisions, by giving us a single comparable value for performance across our 
potential models.

### I. Understanding our Dataset

<img src="image2.jpg">

Our dataset is about houses in Iowa! According to the Kaggle webpage, the competition is 
as follows:

Ask a home buyer to describe their dream house, and they probably won’t begin with the 
height of the basement ceiling or the proximity to an east-west railroad. But this 
playground competition’s dataset proves that much more influences price negotiations than 
the number of bedrooms or a white-picket fence. With 79 explanatory variables describing 
(almost) every aspect of residential homes in Ames, Iowa, this competition challenges you 
to predict the final price of each home.

More explicitly, our dataset has 81 columns, or features:

1. SalePrice, our response variable $Y$ that we are trying to predict accurately and 
precisely
2. Id, a simple identification variable
3. 79 explanatory variables $X_k$ that we can use to predict SalePrice. Some of the variables  are categorical, and others are continuous quantitative.

The goal for this project is to create a model that trains on (some of) the 79 explanators from the training set to predict SalePrice well in the test set.

How will we know which explanators to use? We can start with some intuition and research 
into what each column represents by reading the data dictionary in BLANK.

Please take a moment to read over this dictionary, as you will need to have a keen sense 
of these features for the weeks ahead. Can you come up with five features you suspect will 
be important in determining the SalePrice? (Hint: You may have to do some research of your own outside of this data set to determine if certain variables are relevant).

Now that we've talked a bit about this dataset, let's actually take a look at it. The 
first step is to load in the data! We will store this in a pandas dataframe.

In [196]:
import pandas as pd
train = pd.read_csv('DATA/house-prices/train.csv')
test = pd.read_csv('DATA/house-prices/test.csv')

#Let's see what the training dataframe looks like!
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

We have a training dataset and a testing dataset. You will learn more about how these datasets work together in the linear modelling portion of today's lecture. For now, just focus on the training set.

### Questions to Consider:

> How many features (columns) do we have? How many entries (rows)?
> What does a single row represent in our dataset?
> List 3 issues/questions that you see from the dataset. 
Some ideas to get started:
* What does LotShape represent?
* Is it a good thing that we have so many features? How many features are too many features?
* How does a large number of features affect our modeling approach?
> What are some important things we should do for data cleaning and exploration? Some ideas to get started: are all the values in LotFrontage numeric? What about Alley? How should we fix missing/NA values that appear sporadically in some columns? What about columns that are almost entirely full of NA values? Some columns are qualitative strings, whereas others are qualitative numerics - how might this affect our cleaning?

## Data Cleaning

In [197]:
print("Training size",train.shape)
print("Test size",test.shape)

Training size (1460, 81)
Test size (1459, 80)


This tells us that we have  1460  datapoints in the training set, and we would like to predict on  1459  samples. There are  80  total features, and one response variable we would like to predict. However, not all these 'features' are actually useful, such as Id. Thus we need to understand what the actual variables mean. A huge part in data science is actually understanding the variables. Of course it is possible to throw the data into some machine learning model and have it spit out predictions, but without actually understanding what data you are dealing with and how to feed the data into the model, your model is worthless.

Spend a good deal of time reading over the data dictionary. It is located in
BLANK.

### Questions for Understanding:

> There are many categorical variables. What are some possibilities to deal with these variables? (Hint: OneHotEncoder in sklearn)
> Are there any categorical variables that we can convert to numerical/quantitative variables as well? How might we do that?
> Are there any variables that are just irrelevant and we can ignore?

One helpful method is looking at the unique values of a feature.

In [198]:
#We can view the unique values of a given feature
train['Foundation'].unique()

array(['PConc', 'CBlock', 'BrkTil', 'Wood', 'Slab', 'Stone'], dtype=object)

## Dealing with NA Values

In almost all datasets, we will have NA values. These can be a pain to deal with, as there are many viable choices of what to do. First, it is good to see what columns have NA values.

In [199]:
#Sum the number of NA's in each column
train.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
                 ... 
BedroomAbvGr        0
KitchenAbvGr        0
KitchenQual         0
TotRmsAbvGrd        0
Functional          0
Fireplaces          0
FireplaceQu       690
GarageType         81
GarageYrBlt        81
GarageFinish       81
GarageCars          0
GarageArea          0
GarageQual         81
GarageCond         81
PavedDrive

It seems that many of the features have a great deal of NAs. However, this is not necessarily the case. The astute reader will notice that some variables like Alley, PoolQC, and Fence, have NA has an actual value. You can look at the dictionary to see this for yourself. For example, consider Pool Quality:

In [200]:
train['PoolQC'].unique()

array([nan, 'Ex', 'Fa', 'Gd'], dtype=object)

So having NA values is not the usual not available, it can actually be a legitimate value! We need to parse through the data dictionary to see when NA's are actually significant, and when they mean NA.

The features that we found with NA's are: 

Alley, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, FireplaceQu, GarageType, GarageFinish, GarageQual, GarageCond, PoolQC, Fence, Misc Feature.

* Alley: NA for no alley access
* BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2: NA for no basement
* FireplaceQu: NA for no fireplace
* GarageType, GarageFinish, GarageQual, GarageCond: NA for no garage
* PoolQC: NA for no pool
* Fence: NA for no fence
* Misc Feature: NA for no other miscellanous features (i.e. elevator, 2nd garage, shed, tennis Court, other)

The other categories containing many NA values may not be useful to us. How should we go about deleting them?

## Data Cleaning Functions

We provide some functions to help with data cleaning and preprocessing. #####One is used to create one hot encodings of various features, and the latter is for converting a categorical feature into values.

### Dropping Columns

Often, you’ll find that not all the categories of data in a dataset are useful to you. For example, you might find, during this project, that Fireplace Quality and Lot Shape simply do not influence house price. (You need to see for yourself if this is actually true).

If these two categories are not influencing house price, they are effectively not important to you. Therefore, retaining these variables will take up unnecessary space and could even slow down runtime (which would hurt your chance at winning the Kaggle competition).

Pandas provides a convenient way of removing unwanted columns or rows from a DataFrame with the drop() function. Let’s look at a simple example where we drop a number of columns from our houses data frame.

So suppose that you do some research on how house prices are determined, maybe by making a linear model or just googling factors that influence house prices in Ames, Iowa, and you determine that Garage Condition and Basement Condition are just not very important.

We can drop these columns from the training set in the following way:

In [203]:
to_drop = ['GarageCond', 'BsmtCond']

train.drop(to_drop, inplace=True, axis=1)
#Let's check to see that the columns are gone:
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 79 columns]

In [204]:
#Now I will revert the changes in case you want to keep those columns.
train = pd.read_csv('DATA/house-prices/train.csv')
train.head(10)
#See? the columns are back because I reimported the dataframe.

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

### Using `.loc[ ]`

`dataframe.loc[]` is a location-based indexer that we can use to hone in on a specific row or column value. Suppose we just want to hone in on one particular house sale -- let's say the 1457th house in our list. The 1457th house would be indexed at 1456, since we index starting with 0 in python.

In [205]:
train.loc[1456]

Id                  1457
MSSubClass            20
MSZoning              RL
LotFrontage           85
LotArea            13175
Street              Pave
Alley                NaN
LotShape             Reg
LandContour          Lvl
Utilities         AllPub
LotConfig         Inside
LandSlope            Gtl
Neighborhood      NWAmes
Condition1          Norm
Condition2          Norm
BldgType            1Fam
HouseStyle        1Story
OverallQual            6
OverallCond            6
YearBuilt           1978
YearRemodAdd        1988
RoofStyle          Gable
RoofMatl         CompShg
Exterior1st      Plywood
Exterior2nd      Plywood
MasVnrType         Stone
MasVnrArea           119
ExterQual             TA
ExterCond             TA
Foundation        CBlock
                  ...   
BedroomAbvGr           3
KitchenAbvGr           1
KitchenQual           TA
TotRmsAbvGrd           7
Functional          Min1
Fireplaces             2
FireplaceQu           TA
GarageType        Attchd
GarageYrBlt         1978


Suppose we want to observe all categorical values from the 1457th house to the last (1460th) house. We write:

In [206]:
train.loc[1456:]

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature  \
1456         Lvl    AllPub    ...            0    NaN  MnPrv         NaN   
1457         Lvl    AllPub    ...            0    NaN  GdPrv        Shed   
1458         Lvl    AllPub    ...            0    NaN    NaN         NaN   
1459         Lvl    AllPub    ...            0    NaN    NaN         NaN   

     MiscVal MoSold YrSold  SaleType  SaleCondition  SalePrice  
1456       0      2   2010        WD         Normal     210000  
1457    2500      5   2010        WD         Normal     266500  
1458       0      4   2010        WD         Normal     142125  
1459       0      6   2008        WD         Normal     147500  

[4 rows x 81 columns]

And suppose we don't want to parse through everything we printed just to find what the value of a certain category was for the 1457th house. We can write the following to find out what month the 1457th house was sold in.

In [207]:
train.loc[1456, 'MoSold']

2

The house was sold in February.

I introduced the function loc[] to help you in deciding which columns to remove. Since our dataset is so large it may be easier to look at rows one by one in some cases. How you use or do not use loc[] is up to you!

### Renaming Columns and Skipping Rows

Often, the datasets we work with label their columns in inconvenient ways. For example, maybe the column name is too vague or too long. Or maybe the dataframe will not be _rectangular_, meaning there could be unimportant information in the first few rows.

In that case, we’d want to rename columns and skip certain rows so that we can drill down to necessary information with correct and sensible labels. For example, the following screenshot from MS Excel shows a non-rectangular dataset.

<img src="image.png">

In our case, our data is rectangular, but sometimes datasets can be a lot messier to begin with. So we will just focus on renaming columns. 

To rename the columns, we will make use of a DataFrame’s rename() method, which allows you to relabel an axis based on a mapping.

Suppose that we want the 'Quality' variables to have consistent names and want to change the following to match up with the format of 'OverallQual'.  
* Change 'HeatingQC' to 'HeatingQual'
* Change 'FireplaceQu'to 'FireplaceQual'
* Change 'PoolQC' to 'PoolQual'
And suppose we want to shorten a couple names:
* Change 'Neighborhood' to 'Nbhd'
* 'EnclosedPorch' to 'EnclsdPorch'

Since we are only changing a handful of the columns, we do not want to relist all the columns that are not being renamed. We can use the rename function to tweak only the columns we want:

In [209]:
train.rename(index=str, columns={"HeatingQC": "HeatingQual", 
                                 "FireplaceQu": "FireplaceQual",
                                "PoolQC": "PoolQual",
                                "Neighborhood" : "Nbhd",
                                "EnclosedPorch" : "EnclsdPorch"}, inplace=True)

In [66]:
#Let's check that the columns were renamed:
list(train.columns.values)

['Id',
 'MSSubClass',
 'MSZoning',
 'LotFrontage',
 'LotArea',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Nbhd',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'MasVnrArea',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinSF1',
 'BsmtFinType2',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'Heating',
 'HeatingQual',
 'CentralAir',
 'Electrical',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'KitchenQual',
 'TotRmsAbvGrd',
 'Functional',
 'Fireplaces',
 'FireplaceQual',
 'GarageType',
 'GarageYrBlt',
 'GarageFinish',
 'GarageCars',
 'GarageArea',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'WoodDeckSF',
 'OpenPorchSF',
 '

It worked!

If you don't like the changes made, you can revert the changes by using rename again. You can also make new changes to the column names that make more sense to you.

In [ ]:
#Use this format to change any column names as you see fit:
#train.rename(index=str, columns={"": "", "": ""}, inplace=True)

## One Hot Encoding for Unordered Categorical Variables

### Ordered Categorical Variables

Consider we have a feature attendance for a data frame of SUSA members. Let's say there are  3  possible values, Good, Ok, Poor. These inherently have an ordering, Good is better than Okay which is better than Poor. We can assign a numerical value to these, such as  0  for Poor,  1  for Ok, and  2  for Good. This function is named LabelEncoder and is explained below.


### One Hot Encoder

Consider we have a feature committee for a data frame of SUSA members. Let's say for sake of simplicity there are four committees, CX, DC, RP, and WD. There is no inherent ordering to the features, but each member is only in one committee. Thus we can replace this categorical variable with a single boolean for committee. If a member is a part of CX, then they will have a  1  for DC and  0  for all other variables. This function is named OneHotEncoder and is indicated below.

Please read over the following code carefully. This is how to use Label Encoding and One Hot Encoding together to recategorize a variable. 

In particular, we see from the dictionary that Exterior Quality can take on 5 different values. Upon closer look, it appears that no single house was marked as having 'Poor' quality so we can ignore that option:

In [124]:
train['ExterQual'].unique()

array(['Gd', 'TA', 'Ex', 'Fa'], dtype=object)

Now we use OneHotEncoder and LabelEncoder as follows to create 4 columns, each one designating the the quality of the variable. So in this case, we will have a new column for every house that has 'Gd' ExteriorQual, another for 'TA' ExteriorQual, etc. Python naturally orders the different values alphabetically. Hence, the first column will correspond to 'Excellent', then 'Fair', etc.

In addition, when LabelEncoder works throught the variable chosen, it will naturally associate the following numeric values since it begins at 0 and labels alphabetically:
* Ex is 0
* Fa is 1
* Gd is 2
* TA is 3

To start, let's show how python will naturally associate numeric values with strings.

In [176]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
#here we are importing the necessary functions from sklearn

#let's just reassure ourselves that we are using the original training set. 
#If you made changes to the column names or removed columns already, then
#you may need to comment the next line out.
train = pd.read_csv('DATA/house-prices/train.csv')

#We let the function LabelEncoder associate a value, 0, 1, 2, or 3, 
#with each distinct value that ExterQual takes on. We also create
#a new column that gives the numeric value associated with the 
#value that ExterQual took on for each row and concatenate it with
#the current data set.
enc = LabelEncoder()
train['ExteriorQuality'] = enc.fit_transform(train.ExterQual)

#Now let's print what we have to see that our last column indicates
#ExterQual numerically:
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities       ...        PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub       ...           NaN    NaN         NaN       0   
1         Lvl    AllPub       ...           NaN    NaN         NaN       0   
2         Lvl    AllPub       ...           NaN    NaN         NaN       0   
3         Lvl    AllPub       ...           NaN    NaN         NaN       0   
4         Lvl    AllPub       ...           NaN    NaN         NaN       0   
5         Lvl    AllPub       ...           NaN  MnPrv        Shed     700   
6         Lvl    AllPub       ...           NaN    NaN         NaN       0   
7         Lvl    AllPub       ...           NaN    NaN        Shed     350   
8         Lvl    AllPub       ...           NaN    NaN         NaN       0   
9         Lvl    AllPub       ...           NaN    NaN         NaN       0   

  MoSold YrSold SaleType  SaleCondition  SalePrice  ExteriorQuality  
0      2   2008       WD         Normal     208500                2  
1      5   2007       WD         Normal     181500                3  
2      9   2008       WD         Normal     223500                2  
3      2   2006       WD        Abnorml     140000                3  
4     12   2008       WD         Normal     250000                2  
5     10   2009       WD         Normal     143000                3  
6      8   2007       WD         Normal     307000                2  
7     11   2009       WD         Normal     200000                3  
8      4   2008       WD        Abnorml     129900                3  
9      1   2008       WD         Normal     118000                3  

[10 rows x 82 columns]

As shown in the table above, we indexed each attribute of the category for Exterior Quality using LabelEncoder. We effectively made a categorical variable into a quantitative one.

Now let's complete what we had set out to do -- making individual columns for each possible value of 'ExterQual':

In [185]:
#First let's repeat what was done above.
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder

train = pd.read_csv('DATA/house-prices/train.csv')

#Since we do not desire to add another quantitative column as we did
#above, I labeled our transformed vector 'alpha' and did not attach
#it to our dataset.
enc = LabelEncoder()
alpha = enc.fit_transform(train.ExterQual)
#Now we continue from above to make different columns for each possible
#attribute of the category 'ExterQual'.

#We let the function OneHotEncoder reshape the data because it was
#made into an array separated by value of ExterQual. We are making 
#a column that returns 1 if the feature from the column ExterQual is
#found to be associated with that row.
ohe = OneHotEncoder()
X = ohe.fit_transform(alpha.reshape(-1,1)).toarray()

#Now we label the new columns in a way that is understandable to the viewer
#and we concatenate the columns we created above with our dataset.
trainOneHot = pd.DataFrame(X, columns = ['Excellent ExterQual', 
                                         'Fair ExterQual', 
                                         'Good ExterQual', 
                                         'Average/Typical ExterQual'])
train = pd.concat([train, trainOneHot], axis=1)

#Now we see how this looks!
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities            ...             MiscVal MoSold YrSold  \
0         Lvl    AllPub            ...                   0      2   2008   
1         Lvl    AllPub            ...                   0      5   2007   
2         Lvl    AllPub            ...                   0      9   2008   
3         Lvl    AllPub            ...                   0      2   2006   
4         Lvl    AllPub            ...                   0     12   2008   
5         Lvl    AllPub            ...                 700     10   2009   
6         Lvl    AllPub            ...                   0      8   2007   
7         Lvl    AllPub            ...                 350     11   2009   
8         Lvl    AllPub            ...                   0      4   2008   
9         Lvl    AllPub            ...                   0      1   2008   

  SaleType SaleCondition SalePrice Excellent ExterQual  Fair ExterQual  \
0       WD        Normal    208500                 0.0             0.0   
1       WD        Normal    181500                 0.0             0.0   
2       WD        Normal    223500                 0.0             0.0   
3       WD       Abnorml    140000                 0.0             0.0   
4       WD        Normal    250000                 0.0             0.0   
5       WD        Normal    143000                 0.0             0.0   
6       WD        Normal    307000                 0.0             0.0   
7       WD        Normal    200000                 0.0             0.0   
8       WD       Abnorml    129900                 0.0             0.0   
9       WD        Normal    118000                 0.0             0.0   

   Good ExterQual  Average/Typical ExterQual  
0             1.0                        0.0  
1             0.0                        1.0  
2             1.0                        0.0  
3             0.0                        1.0  
4             1.0                        0.0  
5             0.0                        1.0  
6             1.0                        0.0  
7             0.0                        1.0  
8             0.0                        1.0  
9             0.0                        1.0  

[10 rows x 85 columns]

As we can see above, OneHotEncoder works to replace one column with several, each new column representing whether or not a given row will have that specific attribute of the variable associated with it. We went further than LabelEncoder because we not only made our variable quantitative, but we separated it into several variables.

Use the above two functions to clean the dataset! This may take a while, but doing a good job should take a while. Decide what variables are not worth keeping, decide what categorical features need to be changed, and how they should be changed. Consider how to deal with NA values, and keep all these commands together. We would recommend to save the final cleaned file as a csv so that you may easily reopen and send it, and also keep all the commands together neatly in the code block below.

Some recommendations:

* Remove features that do not seem important.
* Convert all features into quantitative values
* While cleaning, keep in mind some features that you feel are very helpful.

In [ ]:
# DATA CLEANING
# To start off data cleaning
clean = train.copy()

# ====================== DO NOT INCLUDE IN FINAL WORKBOOK ========================
# ================================================================================

    
# ================================================================================
# ================================================================================

# You will be selecting your important features explicitly later. 
# But is there anything you want to get rid of explicitly now?
# Use the following code on any factors you would like to remove. 
# For example, 'Id' is just indexing the number of the house. 
unimportantFeatures = ['Id']
for feat in unimportantFeatures:
    clean = clean.drop(feat, axis=1)

# ADD MORE FEATURES?

clean.head(10)

In [ ]:
To save the dataframe as csv, run this line of code.

In [ ]:
# Save to csv
clean.to_csv('DATA/house-prices/train_cleaned.csv')

Now let's move on toward Data Visualization & Linear Modelling.